### PR TITLE
Dash previews: populate crawler queue from SQL query

### PR DESCRIPTION
**Pull request by ArturWierzbicki**
**What this PR does / why we need it**:

Continuation of https://github.com/grafana/grafana/pull/43509

1. Store thumbnails in the database instead of file system
2. Use SQL query to find dashboards with stale thumbnails

**Which issue(s) this PR fixes**:

Part of https://github.com/grafana/grafana/issues/44449

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/dashboard/components/DashboardSettings/PreviewSettings.tsx [hunk 1]
```diff
@@ -50,6 +55,18 @@ export class PreviewSettings extends PureComponent<Props, State> {
             </tr>
           </thead>
           <tbody>
+            <tr>
+              <td>
+                <Button type="button" variant="primary" onClick={this.markAsStale(false)} fill="outline">
+                  Mark as stale
```
Comment (FZambia) [inline public/app/features/dashboard/components/DashboardSettings/PreviewSettings.tsx:61]: Duplicate text on 2 buttons?

**Review by FZambia — COMMENTED** [review]
Looks like I need to understand better how things work, added some comments, hope to review again after reading previews design doc.

File: pkg/models/dashboard_thumbs.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	case string(ThumbnailKindDefault):
+		return ThumbnailKindDefault, nil
+	}
+	return ThumbnailKindDefault, errors.New("unknown thumbnail kind " + str)
+}
+
+// A DashboardThumbnail includes all metadata for a dashboard thumbnail
+type DashboardThumbnail struct {
+	Id               int64          `json:"id"`
+	DashboardId      int64          `json:"dashboardId"`
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/models/dashboard_thumbs.go:60]: as discussed it looks like we don't need it - can't seem to find a use case for it, but maybe @ryantxu has some idea

[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/models/dashboard_thumbs.go:60]: I'm not sure it is necessary -- the "canView" check works with the dashboard query, and can always be joined by dashboardId to the dashboard table.

File: pkg/models/dashboards.go [hunk 3]
```diff
@@ -317,6 +322,11 @@ func GetDashboardUrl(uid string, slug string) string {
 	return fmt.Sprintf("%s/d/%s/%s", setting.AppSubUrl, uid, slug)
 }
 
+// GetKioskModeDashboardUrl returns the HTML url for a dashboard in kiosk mode.
+func GetKioskModeDashboardUrl(uid string, slug string) string {
```
Comment (ryantxu) [inline pkg/models/dashboards.go:326]: can we avoid public function?  seems like this is not necessary in the models package

File: pkg/services/sqlstore/dashboard_thumbs.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+)
+
+func (ss *SQLStore) GetThumbnail(query *models.GetDashboardThumbnailCommand) (*models.DashboardThumbnail, error) {
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/sqlstore/dashboard_thumbs.go:11]: added

File: pkg/services/sqlstore/migrations/dashboard_thumbs_mig.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package migrations
+
+import "github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+
+func addDashboardThumbsMigrations(mg *migrator.Migrator) {
+	dashThumbs := migrator.Table{
+		Name: "dashboard_thumbnail",
+		Columns: []*migrator.Column{
+			{Name: "id", Type: migrator.DB_BigInt, IsPrimaryKey: true, IsAutoIncrement: true},
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/sqlstore/migrations/dashboard_thumbs_mig.go:9]: same as https://github.com/grafana/grafana/pull/44083#discussion_r793598825

File: pkg/models/dashboard_thumbs.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	// ThumbnailKindDefault is a small 320x240 preview
+	ThumbnailKindDefault ThumbnailKind = "thumb"
+
+	// unsupported for now
+	// - ThumbnailKindLarge ThumbnailKind = "large"
+	// - ThumbnailKindTall ThumbnailKind = "tall"
+)
+
+const (
+	ThumbnailStateDefault ThumbnailState = "default"
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/models/dashboard_thumbs.go:24]: done

File: public/app/features/dashboard/components/DashboardSettings/PreviewSettings.tsx (hunk 1 above)
Comment (ArturWierzbicki) [reply public/app/features/dashboard/components/DashboardSettings/PreviewSettings.tsx:61]: Same text for two different themes ;) The current admin UI is very basic and won't be included in the mvp, I'll disable it in a separate PR https://github.com/grafana/grafana/issues/44449

File: pkg/models/dashboards.go (hunk 3 above)
Comment (ArturWierzbicki) [reply pkg/models/dashboards.go:326]: I mostly follow the approach of grouping together functions/modules that have the same or a similar reason to change. This function constructs a webapp URL pointing at a dashboard, like many other functions in this file. 

Our backend changes would be limited to just a single file if we ever changed the dashboard URL structure, which seems like a good thing

File: pkg/services/sqlstore/dashboard_thumbs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/sqlstore/dashboard_thumbs.go:11]: the tests fail to run in CI due to missing migration script as its hidden behind a feature flag, trying to fix it now

File: pkg/services/thumbs/repo.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	return repo
+}
+
+type sqlThumbnailRepository struct {
+	store *sqlstore.SQLStore
+}
+
+func (r *sqlThumbnailRepository) saveFromFile(filePath string, meta models.DashboardThumbnailMeta, dashboardVersion int) (int64, error) {
+	file, err := os.Open(filepath.Clean(filePath))
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/thumbs/repo.go]: done

File: pkg/services/sqlstore/sqlstore.go [hunk 8]
```diff
@@ -497,7 +501,15 @@ func initTestDB(migration registry.DatabaseMigrator, opts ...InitTestDBOpt) (*SQ
 
 		// set test db config
 		cfg := setting.NewCfg()
-		cfg.IsFeatureToggleEnabled = func(key string) bool { return false }
+		cfg.IsFeatureToggleEnabled = func(requestedFeature string) bool {
```
Comment (ArturWierzbicki) [inline pkg/services/sqlstore/sqlstore.go:504]: we are reusing the same store in different tests so if we have a migration hidden behind a feature flag we have to enable it beforehand. not sure if that's the best way of doing it @FZambia @ryantxu

**Review by papagian — CHANGES_REQUESTED** [review]

**Review by torkelo — COMMENTED** [review]
Just a quick note, in a HA setting where you have potentially many Grafana servers behind a load balancer. Are we using the db lock service to prevent multiple crawlers from all doing the same thing?

File: pkg/services/thumbs/service.go [hunk 9]
```diff
@@ -104,36 +95,75 @@ func (hs *thumbService) parseImageReq(c *models.ReqContext, checkSave bool) *pre
...
+	var body = &updateThumbnailStateRequest{}
+
+	err := web.Bind(c.Req, body)
+	if err != nil {
+		tlog.Error("Error parsing update thumbnail state request", "dashboardUid", req.UID, "err", err.Error())
+		c.JSON(500, map[string]string{"dashboardUID": req.UID, "error": "unknown"})
+		return
+	}
+
+	newState, err := models.ParseThumbnailState(body.State)
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/thumbs/service.go]: changed :heavy_check_mark: 
image

File: pkg/models/dashboard_thumbs.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+//
+// Commands
+//
+
+// DashboardThumbnailMeta uniquely identifies a thumbnail; a natural key
+type DashboardThumbnailMeta struct {
+	DashboardUID string
+	PanelID      int64
+	Kind         ThumbnailKind
+	Theme        Theme
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/models/dashboard_thumbs.go]: good catch! thanks, changed

File: pkg/models/theme.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package models
+
+import "errors"
+
+type Theme string
+
+const (
+	ThemeLight Theme = "light"
+	ThemeDark  Theme = "dark"
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/models/theme.go:9]: created enterprise PR

**Comment by ArturWierzbicki** [comment]
> Just a quick note, in a HA setting where you have potentially many Grafana servers behind a load balancer. Are we using the db lock service to prevent multiple crawlers from all doing the same thing?

@torkelo yeah, that's the idea - will add it in a subsequent PR that runs the crawler as a background service

File: pkg/services/thumbs/crawler.go [hunk 12]
```diff
@@ -83,61 +77,31 @@ func (r *simpleCrawler) broadcastStatus() {
...
-	return &previewResponse{
-		Code: 202,
-		Path: p,
-	}
-}
-
-func (r *simpleCrawler) Start(c *models.ReqContext, mode CrawlerMode, theme rendering.Theme) (crawlStatus, error) {
-	if r.status.State == "running" {
+func (r *simpleCrawler) Start(c *models.ReqContext, mode CrawlerMode, theme models.Theme, thumbnailKind models.ThumbnailKind) (crawlStatus, error) {
+	if r.status.State == running {
```
Comment (papagian) [inline pkg/services/thumbs/crawler.go:81]: I think this check should be locked. what about using `r.Status()` instead?

File: pkg/services/thumbs/repo.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return repo
+}
+
+type sqlThumbnailRepository struct {
+	store *sqlstore.SQLStore
+}
+
+func (r *sqlThumbnailRepository) saveFromFile(filePath string, meta models.DashboardThumbnailMeta, dashboardVersion int) (int64, error) {
+	content, err := os.ReadFile(filepath.Clean(filePath))
```
Comment (papagian) [inline pkg/services/thumbs/repo.go:24]: I think that it's safer to make sure that `filepath` is rooted, given rule 4 of [filepath.Clean()](https://pkg.go.dev/path/filepath#Clean) if the provided path is relative (i.e. not rooted) any occurrences of .. will remain in the resulting path.

File: pkg/services/thumbs/crawler.go (hunk 12 above)
Comment (ArturWierzbicki) [reply pkg/services/thumbs/crawler.go:81]: yeah, I have already changed it in a different PR that targets this branch: https://github.com/grafana/grafana/pull/44891

File: pkg/services/sqlstore/dashboard_thumbs.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+		existing.State = cmd.State
+		_, err = sess.ID(existing.Id).Update(existing)
+		return err
+	})
+
+	return err
+}
+
+func (ss *SQLStore) FindDashboardsWithStaleThumbnails(cmd *models.FindDashboardsWithStaleThumbnailsCommand) ([]*models.DashboardWithStaleThumbnail, error) {
+	err := ss.WithTransactionalDbSession(context.Background(), func(sess *DBSession) error {
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/sqlstore/dashboard_thumbs.go]: done!

File: pkg/services/sqlstore/dashboard_thumbs.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+)
+
+func (ss *SQLStore) GetThumbnail(query *models.GetDashboardThumbnailCommand) (*models.DashboardThumbnail, error) {
+	err := ss.WithTransactionalDbSession(context.Background(), func(sess *DBSession) error {
```
[the remark this answers is not in the document]
Comment (ArturWierzbicki) [reply pkg/services/sqlstore/dashboard_thumbs.go]: done!

File: pkg/services/thumbs/repo.go (hunk 13 above)
Comment (ArturWierzbicki) [reply pkg/services/thumbs/repo.go:27]: added a comment explaining the source of the `filepath` variable

Comment (ArturWierzbicki) [reply pkg/services/thumbs/repo.go:27]: also added `isAbs()` validation

**Review by papagian — APPROVED** [review]
LGTM

**Review by FZambia — APPROVED** [review]
LGTM

**Comment by ArturWierzbicki** [comment]
:rocket:

**Pull request merged**